lib/grandpa/message_handler.go

```diff
@@ -280,21 +302,54 @@ func (h *MessageHandler) verifyCommitMessageJustification(fm *CommitMessage) err
 }
 
 func (h *MessageHandler) verifyPreVoteJustification(msg *CatchUpResponse) (common.Hash, error) {
+	voters := make(map[ed25519.PublicKeyBytes]map[common.Hash]int, len(msg.PreVoteJustification))
+	eqVotesByHash := make(map[common.Hash]map[ed25519.PublicKeyBytes]struct{})
+
+	// identify equivocatory votes by hash
+	for _, justification := range msg.PreVoteJustification {
+		hashsToCount, ok := voters[justification.AuthorityID]
+		if !ok {
+			hashsToCount = make(map[common.Hash]int)
+		}
+
+		hashsToCount[justification.Vote.Hash]++
+		voters[justification.AuthorityID] = hashsToCount
+
+		if hashsToCount[justification.Vote.Hash] > 1 {
+			pubKeysOnHash, ok := eqVotesByHash[justification.Vote.Hash]
+			if !ok {
+				pubKeysOnHash = make(map[ed25519.PublicKeyBytes]struct{})
+			}
+
+			pubKeysOnHash[justification.AuthorityID] = struct{}{}
+			eqVotesByHash[justification.Vote.Hash] = pubKeysOnHash
+		}
+	}
+
 	// verify pre-vote justification, returning the pre-voted block if there is one
 	votes := make(map[common.Hash]uint64)
+	for _, justification := range msg.PreVoteJustification {
+		// if the current voter is on equivocatory map then ignore the vote
+		eqVotersOnHash, ok := eqVotesByHash[justification.Vote.Hash]
+		if ok {
+			_, ok = eqVotersOnHash[justification.AuthorityID]
+			if ok {
+				continue
+			}
+		}
 
-	for _, just := range msg.PreVoteJustification {
-		err := h.verifyJustification(&just, msg.Round, msg.SetID, prevote) //nolint
+		err := h.verifyJustification(&justification, msg.Round, msg.SetID, prevote) //nolint
 		if err != nil {
 			continue
 		}
 
-		votes[just.Vote.Hash]++
+		votes[justification.Vote.Hash]++
 	}
 
 	var prevote common.Hash
 	for hash, count := range votes {
-		if count >= h.grandpa.state.threshold() {
+		eqvOnHash := eqVotesByHash[hash]
+		if count+uint64(len(eqvOnHash)) >= h.grandpa.state.threshold() {
```

Comment:
nit maybe rename to just `equivocatoryVotes`? In the inner for loop local scope, we can see it's the votes for that hash, no need to re-mention it I think.

```suggestion
		equivocatoryVotes := eqVotesByHash[hash]
		if count+uint64(len(equivocatoryVotes)) >= h.grandpa.state.threshold() {
```